### PR TITLE
[Quantization] Add int8 precision for bias of Conv and FC + quantization fixes

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary**
This PR contains the following changes:

- Quantization fix 1: use target precision for bias when computing the quantization profile
The precision used for computing the quantization profile was NOT specialized for bias (see generateNodeQuantizationInfos() in Quantization.cpp). This meant that for bias tensors the quantization profile was not computed with the target precision, rendering it useless (and most probably the reason why biasScale was forced to inputScale * filterScale unconditionally). By solving
this, the quantization params of bias can be used. Also the bias will be quantized to "true" int32 precision (previously the bias scale was manipulated such that only a couple of bits from the entire int32 bitwidth were used).

- Quantization fix 2: manipulation of biasScale.
When forcing the scale of bias to be biasScale = inputScale * filterScale we must verify that the resulting scale is higher than the "true" biasScale (the one from the profile). Otherwise saturations will occur.

- Quantization fix 3: int32 quantization corner case.
After bias is quantized to int32, it will be used in arithmetic operations as (int32)(bias - biasOffset).
To avoid overflow for the bias during subtraction, the bias dynamic range must be int31 (or even less due to other considerations).

- Add Int8 bias precision for Convolution and FullyConnected (CPU backend).
From our experience we concluded that Int8 bias provides similar performance with Int32 bias.
Also Int8 bias allows specialization of the backend kernels to some hand-optimized open source libraries for MCUs (which are commonly implemented with same precision for input/filter/output/bias).

- Add int8/int16 bias precision for Convolution and FullyConnected (Interpreter backend).
This also tests that quantization works with mixed precisions (Int8 with Int8Bias, Int8 with Int32 bias, Int16 with Int16 bias, Int16 with Int32 bias). It also exposed the problems related to "Quantization fix 2".

- Extra unit tests:
When quantizing a float scalar (max == min) the quantize->dequantize pair should preserve the precision

Test Plan
- Unit tests were added for all the features/fixes.

### CODE REVIEW

**Comment by beicy** [comment]
Thanks for working on it. The reason we use int32 for bias is because the partial sum of the matrix-matrix multiplication is accumulated into int32, so int32 bias can be added to the int32 partial sum for better accuracy (i.e. int8 bias caused accuracy drop based on our experiment). The bias size is smaller comparing to others, so using int32 is acceptable.

**Comment by mciprian13** [comment]
The reason we had to add this functionality was to specialize some of the libjit functions to optimized kernel implementations we have no control over (binaries and not source code). Another case is when we have an accelerator which is capable to compute NN functions through RPC calls (again no control over the features - data requirements). We found that some of these libraries/mechanisms implemented the quantized convolution and fully connected using int8 bias (instead of int32). For this we had to make sure Glow is capable to quantize and prepare the bias data as int8.

Of course I don`t want to argue whether int8 or int32 is the better choice (of course int32) but I like to think of Glow as a framework/infrastructure flexible enough to allow the integration of target specific implementations of kernels.

Please mind that this PR also contains some fixes (e.g. in the quantization params where always computed using int8 and not the target precision) plus some extra unit tests.

You might consider that this PR is not of extreme importance to you, but please consider it is important for us :).

Thanks!

**Comment by beicy** [comment]
Thanks a lot for the details!

**Comment by beicy** [comment]
@mciprian13 any updates on CLA ? After that, the PR can be approved. Thanks!

File: tests/unittests/BackendTestUtils.cpp [hunk 1]
```diff
@@ -75,9 +75,9 @@ static Placeholder *createQuantizedPlaceholder(Module &mod,
 
 /// Clone, profile, and run \p origF given the \p bindings and \p EE. \returns
 /// the quantization parameters from the profile given the specified \p schema.
```
Comment (jfix71) [inline tests/unittests/BackendTestUtils.cpp]: nit: Can you update doxygen?

File: lib/Backends/CPU/libjit/libjit_matmul.cpp [hunk 2]
```diff
@@ -308,35 +343,31 @@ void libjit_matmul_i8(int8_t *outW, const int8_t *lhsW, const int8_t *rhsW,
...
+    const int32_t *biasPost, const int32_t *biasScale, const int32_t *outPre,
+    const int32_t *outPost, const int32_t *outScale, const size_t *outWdims,
+    const size_t *inWdims, const size_t *weightsWdims, const size_t *biasWdims,
+    size_t rowNum, int32_t outOffset, int32_t inOffset, int32_t biasOffset) {
+  libjit_rowwise_quantized_fc_generic<int8_t, int8_t>(
+      outW, inW, weightsW, biasW, weightsOffsets, biasPre, biasPost, biasScale,
+      outPre, outPost, outScale, outWdims, inWdims, weightsWdims, biasWdims,
+      rowNum, outOffset, inOffset, biasOffset);
 }
+}
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Backends/CPU/libjit/libjit_matmul.cpp]: Done

File: lib/Backends/Interpreter/Interpreter.cpp [hunk 3]
```diff
@@ -161,7 +165,9 @@ bool Interpreter::isOpSupported(const NodeInfo &NI) const {
     return NI.allInputsAndOutputsHaveSameElemKind(
                {ElemKind::Int8QTy, ElemKind::Int16QTy},
                {ConvolutionNode::BiasIdx}) &&
-           (NI.getInElemTy(ConvolutionNode::BiasIdx) == ElemKind::Int32QTy);
+           (NI.getInElemTy(ConvolutionNode::BiasIdx) == ElemKind::Int8QTy ||
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Backends/Interpreter/Interpreter.cpp]: Done.

File: lib/Backends/CPU/libjit/libjit_conv.cpp [hunk 4]
```diff
@@ -419,103 +518,35 @@ void libjit_convolution_f(float *outW, const float *inW, const float *filterW,
...
-                    int32_t in = inW[inIdx + fd] - inOffset;
-                    for (unsigned i = 4; i < MIN(8, depthUnroll); i++) {
-                      sum[i] += (filterW[filterIdx + (sliceSize * i) + fd] -
-                                 filterOffset) *
-                                in;
-                    }
-                  }
-              }
-            }
+void libjit_convolution_i32_i8(
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Backends/CPU/libjit/libjit_conv.cpp]: Done.

File: tests/unittests/BackendTestUtils.cpp (hunk 1 above)
Comment (mciprian13) [reply tests/unittests/BackendTestUtils.cpp]: Done.

File: lib/Backends/CPU/libjit/libjit_matmul.cpp [hunk 5]
```diff
@@ -308,35 +343,31 @@ void libjit_matmul_i8(int8_t *outW, const int8_t *lhsW, const int8_t *rhsW,
   }
 }
 
-void libjit_rowwise_quantized_fc_i8(
+/// Rowwise quantized FullyConnected with int8 precision and int32 bias.
+void libjit_rowwise_quantized_fc_i32_i8(
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Backends/CPU/libjit/libjit_matmul.cpp]: Done

**Comment by mciprian13** [comment]
Regarding the CLA, it seems the policy of my company does not allow signing click-through agreements. It seems some request for a more formal CLA has been made. Still waiting for this process to be over.

**Comment by mciprian13** [comment]
The CLA has been signed. I think I also done the review items you requested (please tell me if I forgot something). Let`s have this landed.
Thanks!

File: lib/Quantization/Quantization.cpp [hunk 6]
```diff
@@ -889,11 +912,49 @@ generateNodeQuantizationInfos(PlaceholderBindings &bindings, const Function *F,
...
+      // TODO: For better clarity and to remove logic duplication this code
+      // should be coupled tighter with the logic used in FunctionQuantizer.
+      NodeValue profNode = QPN->getInput();
+      for (const auto &use : profNode.getUsers()) {
+        const auto *user = use.getUser();
+        if ((user->getKind() == glow::Kinded::Kind::ConvolutionNodeKind) &&
+            (user->getNthInput(ConvolutionNode::BiasIdx) == profNode)) {
+          // Found bias for ConvolutionNode.
+          qPrec = quantizationPrecisionBias;
+        }
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Quantization/Quantization.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 7]
```diff
@@ -3779,6 +3779,38 @@ TEST_P(OperatorStatelessTest, FP16ConvolutionDepth8) {
                             parCloneCountOpt);
 }
 
+TEST_P(OperatorStatelessTest, ConvolutionDepth10_Int8_BiasInt8) {
+  ENABLED_BACKENDS(Interpreter, CPU);
+  compareAgainstInterpreter(
+      getBackendName(), createAndInitConvDepthTest<10>, ElemKind::FloatTy,
+      ElemKind::Int8QTy, 0.03f, parCloneCountOpt, false,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 8]
```diff
@@ -3779,6 +3779,38 @@ TEST_P(OperatorStatelessTest, FP16ConvolutionDepth8) {
...
+      quantization::Schema::Asymmetric, ElemKind::Int8QTy);
+}
+
+TEST_P(OperatorStatelessTest, ConvolutionDepth10_Int8_BiasInt32) {
+  ENABLED_BACKENDS(Interpreter, CPU);
+  compareAgainstInterpreter(
+      getBackendName(), createAndInitConvDepthTest<10>, ElemKind::FloatTy,
+      ElemKind::Int8QTy, 0.03f, parCloneCountOpt, false,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 9]
```diff
@@ -3779,6 +3779,38 @@ TEST_P(OperatorStatelessTest, FP16ConvolutionDepth8) {
...
+      quantization::Schema::Asymmetric, ElemKind::Int32QTy);
+}
+
+TEST_P(OperatorStatelessTest, ConvolutionDepth10_Int16_BiasInt16) {
+  ENABLED_BACKENDS(Interpreter);
+  compareAgainstInterpreter(
+      getBackendName(), createAndInitConvDepthTest<10>, ElemKind::FloatTy,
+      ElemKind::Int16QTy, 0.0003f, parCloneCountOpt, false,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 10]
```diff
@@ -3779,6 +3779,38 @@ TEST_P(OperatorStatelessTest, FP16ConvolutionDepth8) {
...
+      quantization::Schema::Asymmetric, ElemKind::Int16QTy);
+}
+
+TEST_P(OperatorStatelessTest, ConvolutionDepth10_Int16_BiasInt32) {
+  ENABLED_BACKENDS(Interpreter);
+  compareAgainstInterpreter(
+      getBackendName(), createAndInitConvDepthTest<10>, ElemKind::FloatTy,
+      ElemKind::Int16QTy, 0.0003f, parCloneCountOpt, false,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 11]
```diff
@@ -3878,6 +3910,42 @@ TEST_P(OperatorStatelessTest, FC_Float16) {
                             parCloneCountOpt);
 }
 
+/// Test Int8 FullyConnected with Int8 bias.
+TEST_P(OperatorStatelessTest, FullyConnected_Int8_BiasInt8) {
+  ENABLED_BACKENDS(Interpreter, CPU);
+  compareAgainstInterpreter(
+      getBackendName(), createAndInitBasicFCTest, ElemKind::FloatTy,
+      ElemKind::Int8QTy, 0.05f, parCloneCountOpt, false,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 12]
```diff
@@ -3878,6 +3910,42 @@ TEST_P(OperatorStatelessTest, FC_Float16) {
...
+      quantization::Schema::Asymmetric, ElemKind::Int8QTy);
+}
+
+/// Test Int8 FullyConnected with Int32 bias.
+TEST_P(OperatorStatelessTest, FullyConnected_Int8_BiasInt32) {
+  ENABLED_BACKENDS(Interpreter, CPU);
+  compareAgainstInterpreter(
+      getBackendName(), createAndInitBasicFCTest, ElemKind::FloatTy,
+      ElemKind::Int8QTy, 0.05f, parCloneCountOpt, false,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 13]
```diff
@@ -3878,6 +3910,42 @@ TEST_P(OperatorStatelessTest, FC_Float16) {
...
+      quantization::Schema::Asymmetric, ElemKind::Int32QTy);
+}
+
+/// Test Int16 FullyConnected with Int16 bias.
+TEST_P(OperatorStatelessTest, FullyConnected_Int16_BiasInt16) {
+  ENABLED_BACKENDS(Interpreter);
+  compareAgainstInterpreter(
+      getBackendName(), createAndInitBasicFCTest, ElemKind::FloatTy,
+      ElemKind::Int16QTy, 0.0005f, parCloneCountOpt, false,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 14]
```diff
@@ -3878,6 +3910,42 @@ TEST_P(OperatorStatelessTest, FC_Float16) {
...
+      quantization::Schema::Asymmetric, ElemKind::Int16QTy);
+}
+
+/// Test Int16 FullyConnected with Int32 bias.
+TEST_P(OperatorStatelessTest, FullyConnected_Int16_BiasInt32) {
+  ENABLED_BACKENDS(Interpreter);
+  compareAgainstInterpreter(
+      getBackendName(), createAndInitBasicFCTest, ElemKind::FloatTy,
+      ElemKind::Int16QTy, 0.0005f, parCloneCountOpt, false,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 15]
```diff
@@ -6297,6 +6365,100 @@ TEST_P(OperatorTest, NonCubicKernelConv3DQuantized) {
   }
 }
 
+/// Test for quantized Convolution3D.
+static void Conv3DQuantizedTest(glow::PlaceholderBindings &bindings,
+                                glow::Module &mod, glow::Function *F,
+                                glow::ExecutionEngine &EE, ElemKind elemKind,
+                                ElemKind biaselemKind) {
+  // Create floating-point network
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 16]
```diff
@@ -6297,6 +6365,100 @@ TEST_P(OperatorTest, NonCubicKernelConv3DQuantized) {
...
+  auto *filter = mod.createPlaceholder(ElemKind::FloatTy, {1, 1, 2, 3, 1},
+                                       "filter", false);
+  auto *bias = mod.createPlaceholder(ElemKind::FloatTy, {1}, "bias", false);
+  auto outTy = mod.uniqueType(ElemKind::FloatTy, {1, 4, 3, 2, 1});
+  Convolution3DNode *conv3d =
+      F->createConv3D("Conv3D", input, filter, bias, outTy, {1, 2, 3},
+                      {1, 1, 1}, {0, 0, 0, 0, 0, 0}, 1);
+  SaveNode *save = F->createSave("save", conv3d);
+
+  // Quantized types
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 17]
```diff
@@ -6297,6 +6365,100 @@ TEST_P(OperatorTest, NonCubicKernelConv3DQuantized) {
...
+  auto inputTQP = quantization::chooseQuantizationParams(
+      -1.0, 1.0, quantization::Schema::Asymmetric, elemKind);
+  auto filterTQP = quantization::chooseQuantizationParams(
+      -1.0, 1.0, quantization::Schema::Asymmetric, elemKind);
+  auto biasTQP = quantization::chooseQuantizationParams(
+      -1.0, 1.0, quantization::Schema::Asymmetric, biaselemKind);
+  auto outputTQP = quantization::chooseQuantizationParams(
+      -4.0, 4.0, quantization::Schema::Asymmetric, elemKind);
+
+  // Create quantized network
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 18]
```diff
@@ -6297,6 +6365,100 @@ TEST_P(OperatorTest, NonCubicKernelConv3DQuantized) {
...
+  QuantizeNode *inputQ = F->createQuantize("inputQ", input, inputQTy);
+  QuantizeNode *filterQ = F->createQuantize("filterQ", filter, filterQTy);
+  QuantizeNode *biasQ = F->createQuantize("biasQ", bias, biasQTy);
+  Convolution3DNode *conv3dQ =
+      F->createConv3D("Conv3DQ", inputQ, filterQ, biasQ, outQTy, {1, 2, 3},
+                      {1, 1, 1}, {0, 0, 0, 0, 0, 0}, 1);
+  DequantizeNode *deQ = F->createDequantize("deQ", conv3dQ);
+  SaveNode *saveQ = F->createSave("saveQ", deQ);
+
+  // Allocate placeholders
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 19]
```diff
@@ -6297,6 +6365,100 @@ TEST_P(OperatorTest, NonCubicKernelConv3DQuantized) {
...
+  bindings.allocate(input)->getHandle().randomize(-1.0, 1.0, mod.getPRNG());
+  bindings.allocate(filter)->getHandle().randomize(-1.0, 1.0, mod.getPRNG());
+  bindings.allocate(bias)->getHandle().randomize(-1.0, 1.0, mod.getPRNG());
+  bindings.allocate(save->getPlaceholder());
+  bindings.allocate(saveQ->getPlaceholder());
+
+  // Run network
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

File: tests/unittests/OperatorTest.cpp [hunk 20]
```diff
@@ -6297,6 +6365,100 @@ TEST_P(OperatorTest, NonCubicKernelConv3DQuantized) {
...
+  ::glow::convertPlaceholdersToConstants(
+      F, bindings, {input, save->getPlaceholder(), saveQ->getPlaceholder()});
+  EE.compile(CompilationMode::Infer);
+  EE.run(bindings);
+
+  // Compare
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tests/unittests/OperatorTest.cpp]: Done.

**Pull request closed without merging**